command/command.go

```diff
@@ -0,0 +1,126 @@
+package command
+
+import (
+	"bufio"
+	"io"
+	"os"
+	"os/exec"
+	"sync"
+)
+
+// ExternalCommand has the properties for invoking exec.Command
+type ExternalCommand struct {
+	Path string
+	Arg  []string
+	Env  []string
+}
+
+type ProgramState struct {
+	Command     *exec.Cmd
+	stdoutPipe  io.ReadCloser
+	stderrPipe  io.ReadCloser
+	stdout      []byte
+	stderr      []byte
+	stdoutmutex sync.Mutex
+	stderrmutex sync.Mutex
+}
+
+// ExecuteProgram invokes an external program
+func ExecuteProgram(command ExternalCommand, errorHandler func(*ProgramState, error) error) (*ProgramState, error) {
+	var err error
+	programState := &ProgramState{}
+	cmd := exec.Command(command.Path, command.Arg...)
+	cmd.Env = append(os.Environ(), command.Env...)
+	programState.Command = cmd
+
+	programState.stdoutPipe, err = cmd.StdoutPipe()
+	if err != nil {
+		return programState, err
+	}
+	programState.stderrPipe, err = cmd.StderrPipe()
+	if err != nil {
+		return programState, err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return programState, err
+	}
+
+	go CloseProgram(programState, errorHandler)
+
+	return programState, err
+}
+
+func CloseProgram(programState *ProgramState, errorHandler func(*ProgramState, error) error) {
+
+	var waitGroup sync.WaitGroup
+	waitGroup.Add(1)
+
+	go func() {
+		errStdout := programState.StdoutListner()
+		if errStdout != nil {
+			errorHandler(programState, errStdout)
+			return
+		}
+		waitGroup.Done()
+	}()
+
+	errStderr := programState.StderrListner()
+	if errStderr != nil {
+		errorHandler(programState, errStderr)
+		return
+	}
+	waitGroup.Wait()
+
+	err := programState.Command.Wait()
+	errorHandler(programState, err)
+
+}
+
+func (p *ProgramState) StdoutListner() error {
+
+	return Listner(p.stdoutPipe, p.stdout, &p.stdoutmutex)
+
+}
+
+func (p *ProgramState) StderrListner() error {
+
+	return Listner(p.stderrPipe, p.stderr, &p.stderrmutex)
+}
+
+func Listner(pipe io.ReadCloser, buffer []byte, mutex *sync.Mutex) error {
+
+	reader := bufio.NewReader(pipe)
+	for {
+
+		buf, err := reader.ReadBytes('\n')
+		mutex.Lock()
+		buffer = append(buffer, buf...)
```

Comment:
This is a bit tricky. So even though slices are references to an underlying array, it is not guaranteed that append with change the same slice. It can also return a new slice. Which means you need to take in a pointer to a slice, and assign to *buffer.